File: src/services/hotelService.ts
import apiClient from './api';
import { Hotel, HotelRoom } from '../types';

export const hotelService = {
  async getAll(): Promise<Hotel[]> {
    const { data } = await apiClient.get<Hotel[]>('/hotels');
    return data;
  },

  async getById(id: string): Promise<Hotel> {
    const { data } = await apiClient.get<Hotel>(`/hotels/${id}`);
    return data;
  },

  async getByDestination(destinationId: string): Promise<Hotel[]> {
    const { data } = await apiClient.get<Hotel[]>(`/hotels/destination/${destinationId}`);
    return data;
  },

  async getByPartner(partnerId: string): Promise<Hotel[]> {
    const { data } = await apiClient.get<Hotel[]>(`/hotels/partner/${partnerId}`);
    return data;
  },

  async create(hotel: Omit<Hotel, 'id'>): Promise<Hotel> {
    const { data } = await apiClient.post<Hotel>('/hotels', hotel);
    return data;
  },

  async update(id: string, hotel: Partial<Hotel>): Promise<Hotel> {
    const { data } = await apiClient.put<Hotel>(`/hotels/${id}`, hotel);
    return data;
  },

  async updateRoomPrice(hotelId: string, roomId: string, price: number): Promise<Hotel> {
    const { data } = await apiClient.patch<Hotel>(`/hotels/${hotelId}/rooms/${roomId}/price`, { price });
    return data;
  },

  async toggleRoomAvailability(hotelId: string, roomId: string): Promise<Hotel> {
    const { data } = await apiClient.patch<Hotel>(`/hotels/${hotelId}/rooms/${roomId}/toggle`);
    return data;
  },

  async addRoom(hotelId: string, room: Omit<HotelRoom, 'id' | 'availableCount'>): Promise<Hotel> {
    const { data } = await apiClient.post<Hotel>(`/hotels/${hotelId}/rooms`, room);
    return data;
  },

  async delete(id: string): Promise<void> {
    await apiClient.delete(`/hotels/${id}`);
  },
};

export default hotelService;

File: backend/src/main/java/com/voyago/service/HotelService.java
package com.voyago.service;

import com.voyago.model.Hotel;
import com.voyago.model.HotelRoom;
import com.voyago.repository.HotelRepository;
import com.voyago.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class HotelService {

    private final HotelRepository hotelRepository;
    private final UserRepository userRepository;

    public List<Hotel> getAll() {
        return hotelRepository.findAll().stream().map(this::syncApprovedPartner).toList();
    }

    public Hotel getById(String id) {
        return hotelRepository.findById(id).map(this::syncApprovedPartner)
                .orElseThrow(() -> new RuntimeException("Hotel not found: " + id));
    }

    public List<Hotel> getByDestination(String destinationId) {
        return hotelRepository.findByDestinationId(destinationId);
    }

    public List<Hotel> getByPartner(String partnerId) {
        return hotelRepository.findByPartnerId(partnerId).stream().map(this::syncApprovedPartner).toList();
    }

    private Hotel syncApprovedPartner(Hotel hotel) {
        if (!"PENDING".equals(hotel.getApprovalStatus()) || hotel.getPartnerId() == null)
            return hotel;
        userRepository.findById(hotel.getPartnerId()).ifPresent(user -> {
            if ("APPROVED".equals(user.getPartnerStatus())) {
                hotel.setApprovalStatus("APPROVED");
                hotel.setStatus("ACTIVE");
                hotelRepository.save(hotel);
            }
        });
        return hotel;
    }

    public Hotel create(Hotel hotel) {
        return hotelRepository.save(hotel);
    }

    public Hotel update(String id, Hotel updated) {
        Hotel existing = getById(id);
        updated.setId(existing.getId());
        return hotelRepository.save(updated);
    }

    public Hotel updateRoomPrice(String hotelId, String roomId, double newPrice) {
        Hotel hotel = getById(hotelId);
        hotel.getRooms().stream()
                .filter(r -> r.getId().equals(roomId))
                .findFirst()
                .ifPresent(r -> r.setPricePerNight(newPrice));
        return hotelRepository.save(hotel);
    }

    public Hotel toggleRoomAvailability(String hotelId, String roomId) {
        Hotel hotel = getById(hotelId);
        hotel.getRooms().stream()
                .filter(r -> r.getId().equals(roomId))
                .findFirst()
                .ifPresent(r -> r.setActive(!r.isActive()));
        return hotelRepository.save(hotel);
    }

    public void delete(String id) {
        hotelRepository.deleteById(id);
    }
}
